### PR TITLE
feat: implement FBS to ANSI/NIST-ITL Type-9 Minutiae Mapping & XML Serialization

### PR DESCRIPTION
**Pull request by dawid200272**
# Pull Request: Implement FBS to ANSI/NIST-ITL Type-9 Minutiae Mapping & XML Serialization

## 📝 Description
This Pull Request delivers a fully tested, standard-compliant pipeline to map, transform, and serialize fingerprint biometric markings from the internal **FBS (Forensic Biometrics Studio)** format into the **ANSI/NIST-ITL 1-2011: UPDATE 2015** standard, specifically adhering to the **INTERPOL v6.00.02 XML implementation profile**.

The implementation is split into four foundational layers:
1. **Core Data Structures & Enums:** Rigid definitions of standard biometric classification categories.
2. **Domain Object Models:** Polymorphic representation of fingerprint features.
3. **Forensic Mapping Engine:** Geometrical coordinate mapping and orientation angle quantization.
4. **Encapsulated XML Serialization:** Generation of human-readable Type-9 (`PackageMinutiaeRecord`) XML blocks.

---

### 1. Core Data Structures & Biometric Domain Models (`feat`)
* **Biometric Standard Enums:** Introduced `InterpolMinutiaeTypes.ts` to establish core dictionary standards. It defines explicit, strict classifications (such as `MinutiaType` representing RidgeEnding, RidgeBifurcation, and Other) directly extracted from the official INTERPOL XML implementation profile reference tables.
* **Domain Object Models Hierarchy:** Created a type-safe class layout to encapsulate fingerprint points: `MinutiaBase`, `Minutia` (standard INCITS points), `CoreMinutia`, and `DeltaMinutia`.
* **Strict Shape Contracts:** Leveraged TypeScript parameter properties and explicit property overrides (`override` keyword) to shield mapped objects against unintended runtime alterations.
* **NIST Specification Cross-Referencing:** Embedded detailed JSDoc documentation linking class fields directly to their corresponding standard NIST data fields (e.g., `Field 09.137 / FMD`, `Field 09.139 / CIN`, `Field 09.140 / DIN`, and `Field 09.141 / ADA`) ensuring hover context inside IDEs.

### 2. Coordinate Mapping & Angular Quantization (`feat` & `refactor`)
* Implemented `MinutiaMapper.ts` to translate internal application markings (`RayMarking`, `PointMarking`) into unified biometric standard primitives.
* Developed custom mathematical conversion logic to map application coordinate spaces and handle standard rotation alignments:
  * **Direction Correction:** Inverts the rotation direction and dynamically shifts the 0-origin axis from *Down* (UI space) to *Right* (NIST standard space).
  * **Theta Quantization:** Quantizes continuous radian values into strict 2-degree integer steps bounded tightly within the $[0, 179]$ step range required by the INCITS 378 specification blocks.
* **Refactoring:** Decoupled the mapping component entirely from the global reactive state (`MarkingTypesStore.state.types`) by utilizing **Dependency Injection (DI)** via method parameters. This isolates the mapping engine and enables reliable standalone unit testing.
* **Simplification:** Optimized the boundary angle wrapping statements using a clean modulo-bound step expression (`% 180`).

### 3. Isolated XML Serializer Engine (`feat` & `refactor`)
* Introduced `MinutiaeXmlSerializer.ts` to fully isolate the generation of `Type-9` logical records from other domain boundaries.
* Renamed and decoupled the legacy writer structures to separate layout builders from processing logic. Sub-element engines (`serializeStandardMinutia`, `serializeCoreLocation`, `serializeDeltaLocation`) are now private and strictly encapsulated.
* Exposed a singular, clean, unified public entry point: `serializeToType9AnsiRecordXml`.
* Programmed dynamic XML block injection supporting multi-angle arrays (primary, secondary, and tertiary orientations) exclusively for fingerprint `DeltaMinutia` markings.
* Replaced manual string concatenation loops with continuous line joins (`.join("\n")`) to guarantee pretty-printed, human-readable XML documents.

### 4. Comprehensive Unit Test Suites (`test`)
* Added 100% test coverage using Jest for both decoupled business layers: `MinutiaMapper.test.ts` and `MinutiaeXmlSerializer.test.ts`.
* Implemented Black-Box validation strategies covering:
  * Real-world operational test vectors extracted from raw operational JSON application files (a sample output JSON file).
  * Critical mathematical boundary conditions (e.g., verifying that raw radians resulting exactly in 180° wrap safely back to 0°).
  * Structural integrity tests proving empty collections generate an empty, schema-compliant XML wrapper containing `<biom:MinutiaeQuantity>0</biom:MinutiaeQuantity>`.
  * Fully documented all test cases using JSDoc syntax.

---

## 📋 Biometric Field Coverage Checklist

| Standard Reference | Technical Description | Implementation Status |
| :--- | :--- | :---: |
| **Field 09.002 (IDC)** | Information Designation Character Identification | ✔ |
| **Field 09.137 (FMD)** | Minutia Identification, Coordinates, Theta Angle, INCITS Category Code, and Minutia Quality | ✔ |
| **Field 09.139 (CIN)** | Fingerprint Pattern Core Location Coordinates & Main Theta Angle | ✔ |
| **Field 09.140 (DIN)** | Fingerprint Pattern Delta Location Coordinates & Primary Theta Angle | ✔ |
| **Field 09.141 (ADA)** | Dynamic Secondary & Tertiary Delta Theta Angles | ✔ |

---

## 🧪 Verification & Test Execution Results

All automated test suites pass cleanly inside the local environment:

```bash

> biometrics-studio@0.6.10 test:unit
> jest ./__tests__/unit --passWithNoTests

 PASS  __tests__/unit/MinutiaeXmlSerializer.test.ts
  MinutiaeXmlSerializer - serializeToType9AnsiRecordXml
    √ should generate a valid base Type-9 XML structure when the minutiae array is empty (11 ms)
    √ should correctly serialize standard INCITS minutiae and reflect accurate quantity (1 ms)
    √ should correctly serialize Core locations (1 ms)
    √ should serialize Delta location with only the primary angle when secondary angles are missing (1 ms)
    √ should dynamically serialize Delta location with multiple angles (secondary and tertiary)
    √ should separate multiple elements with newlines for human-readable formatting
    √ should correctly handle a mixed array containing all minutiae classes simultaneously

# __tests__/unit/autoMarkWithSourceafis.mapping.test.ts results omitted from output

 PASS  __tests__/unit/MinutiaMapper.test.ts
  MinutiaMapper - mapMinutiaFromFbsToAnsiNist
    √ should correctly round coordinates and default angle to 0 for non-ray markings (10 ms)
    √ should map to CoreMinutia class when marking type is 'Core' (1 ms)
    √ should map to DeltaMinutia class when marking type is 'Delta' (1 ms)
    √ should map to standard Minutia and resolve proper MinutiaType for Bifurcation
    √ should map to standard Minutia and resolve proper MinutiaType for RidgeEnding (1 ms)
    √ should fallback to MinutiaType.Other if marking type is unrecognized or missing
    √ should return Minutia with type Other if the marking type exists in store but is unmapped in switch-case (1 ms)
    √ should fallback to quality 0 if minutiaQuality parameter is omitted (7 ms)
    √ should accurately convert app radians to quantized 2-degree Interpol steps (1 ms)
    √ should wrap incitsDegrees from 180 to 0 due to the quantization boundary condition (6 ms)

Test Suites: 3 passed, 3 total
Tests:       20 passed, 20 total
Snapshots:   0 total
Time:        2.815 s
Ran all test suites matching /.\\__tests__\\unit/i.

```

### CODE REVIEW

File: src/lib/ansi-nist-itl/MinutiaMapper.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * 2. Converts the adjusted radian value into raw decimal degrees.
+ * 3. Normalizes the decimal degree value into a standard [0, 360) circle via double modulo operations.
+ * 4. Quantizes the result into standard 2-degree increments [0, 179] matching INCITS 378 / NIST blocks (where 180 wraps to 0).
+ *
+ * @param appRadians - Input angle value represented in radians from the application coordinate system.
+ * @returns Standardized orientation value mapped to 2-degree integer steps within the [0, 179] range.
+ */
+function mapAppRadiansToInterpolDegrees(appRadians: number): number {
+    // 1. Invert the rotation direction and shift the 0-origin from Down to Right
+    const correctedRadians = -appRadians - Math.PI / 2.0;
```
Comment (TheMultii) [inline src/lib/ansi-nist-itl/MinutiaMapper.ts:47]: src/lib/utils/viewport/saveAnsiNistWithDialog.ts:62

why the angle convention conflicts with the two existing conventions?

Comment (dawid200272) [reply src/lib/ansi-nist-itl/MinutiaMapper.ts:47]: In ANSI/NIST-ITL standard documentation [NIST.SP.500-290e3](https://github.com/INTERPOL-Innovation-Centre/ANSI-NIST-XML-ITL-Implementation/blob/master/Documentation/NIST%20Publications/NIST.SP.500-290e3.pdf) for the field 9.137 "M1 finger minutiae data" the angle of minutia is defined as "recorded in units of two degrees. This value shall be nonnegative between 0 and 179, inclusive.". Which is also stated in Interpol's implementation of the standard [NIST INTERPOL standard v6.00.02](https://github.com/INTERPOL-Innovation-Centre/ANSI-NIST-XML-ITL-Implementation/blob/master/Documentation/NIST%20INTERPOL%20standard%20v6.00.02.pdf) for the "Minutia Angle - MAV" (which references the 09.137-D field), the value range is defined as from 0 to 179.

In the document [M1- Finger Minutiae/Image Harmonization](https://www.nist.gov/system/files/documents/2021/03/05/ansi-nist_archived_2007_workshop2_mccabe-m1-harmon.pdf) for the field "Finger Minutiae Data", the angle is also defined as recorded in units of two degrees.

Comment (dawid200272) [reply src/lib/ansi-nist-itl/MinutiaMapper.ts:47]: It looks like the code for saving minutiae data in the XML ANSI/NIST format (`src/lib/utils/viewport/saveAnsiNistWithDialog.ts:62`) is using the traditional NIST format used in the binary format for minutiae data, now used in the legacy fields 09.005 - 09.012 of the ANSI/NIST-ITL standard (2015 update) [NIST.SP.500-290e3](https://github.com/INTERPOL-Innovation-Centre/ANSI-NIST-XML-ITL-Implementation/blob/master/Documentation/NIST%20Publications/NIST.SP.500-290e3.pdf) only for interchange of legacy data.
The Interpol's implementation however deprecated the use of the earlier format and uses only the newer XML-based version.
